### DISCUSSION TITLE
Cannot get custom Alpine directives working with npm modules

### DISCUSSION THREAD

**Discussion by dotfrag** [alpinejs/alpine#4424] [1. Help] [answered]
I can't get custom Alpine directive to work. I'm using rollup, and my issue is probably related to:

> Because these APIs have an impact on how Alpine initializes the page, they must be registered AFTER Alpine is downloaded and available on the page, but BEFORE it has initialized the page itself.

But I have no idea how to go about it.

`rollup.config.mjs`:
```js
import nodeResolve from "@rollup/plugin-node-resolve";
import terser from "@rollup/plugin-terser";

export default {
	input: "static/main.js",
	output: {
		file: "static/bundle.min.js",
		format: "es",
		plugins: [terser()],
	},
	plugins: [nodeResolve()],
	onwarn: (entry, next) => {
		if (
			entry.loc?.file &&
			/htmx\.esm\.js$/.test(entry.loc.file) &&
			/Use of eval in/.test(entry.message)
		)
			return;
		return next(entry);
	},
};
```

`main.js`:
```js
import Alpine from "alpinejs";
import htmx from "htmx.org";

Alpine.directive("uppercase", (el) => {
  el.textContent.toUpperCase();
});

window.Alpine = Alpine;
window.Alpine.start();

htmx.onLoad(() => {
  console.log("htmx loaded");
});
```

`index.html` (truncated):
```html
<html>
<meta>
        <script defer src="{% static 'bundle.min.js' %}"></script>
</meta>

<body>
            <div x-data>
                <span x-uppercase>Hello World!</span>
            </div>
</body>
</html>
```

If I use the `Via a script tag` method, it works fine. Any help is appreciated.

**Comment by ekwoka** [comment]
the above code works fine.

What makes you think it's not working?

**Reply by dotfrag** [reply]
The `<span>` element is not shown in uppercase.

**Reply by ekwoka** [reply]
That's because your code never sets it to be uppercase.

If you did

```js
el.textContent = el.textContent.toUpperCase()
```

You'd get the result you want.

(this should be done in CSS, but I assume this is just a simplistic recreation)

**Reply by dotfrag** [reply]
It's simplistic because I'm setting up a project and just following the [guide](https://alpinejs.dev/advanced/extending).

I didn't get your suggestion sorry, I believe that's what I'm doing already?

**Reply by ekwoka** [reply, marked as answer]
Your code is

```js
Alpine.directive("uppercase", (el) => {
  el.textContent.toUpperCase();
});
```

But the correct code is

```js
Alpine.directive("uppercase", (el) => {
  el.textContent = el.textContent.toUpperCase();
});
```

**Reply by dotfrag** [reply]
This is embarrassing, I can't tell you how much time I spent tinkering with my setup. I didn't realise I had the wrong code..
